e2e/pos_test.go

```diff
@@ -121,3 +127,400 @@ func TestPoS_Unstake(t *testing.T) {
+		signedTx, signErr := signer.SignTx(&types.Transaction{
+			Nonce:    uint64(currentNonce),
+			From:     types.ZeroAddress,
+			To:       &to,
+			GasPrice: gasPrice,
+			Gas:      1000000,
+			Value:    value,
+			V:        1, // it is necessary to encode in rlp
+		}, senderKey)
+
+		if signErr != nil {
+			t.Fatalf("Unable to sign transaction, %v", signErr)
+		}
+
+		currentNonce++
+
+		return signedTx
+	}
+
+	// Test scenario:
+	// User has 10 ETH staked and a balance of 10 ETH
+	// Unstake -> Stake 1 ETH -> Unstake -> Stake 1 ETH...
+	// The code below tests (numTransactions / 2) cycles of Unstake -> Stake 1 ETH
+	// Expected result for tests: Staked: 1 ETH; Balance: ~19 ETH
+
+	oneEth := tests.EthToWei(1)
+	zeroEth := tests.EthToWei(0)
+	for i := 0; i < numTransactions; i++ {
+		var msg *txpoolOp.AddTxnReq
+		if i%2 == 0 {
+			unstakeTxn := generateTx(zeroEth, unstakingContractAddr)
+			msg = &txpoolOp.AddTxnReq{
+				Raw: &any.Any{
+					Value: unstakeTxn.MarshalRLP(),
+				},
+				From: types.ZeroAddress.String(),
+			}
+		} else {
+			stakeTxn := generateTx(oneEth, stakingContractAddr)
+			msg = &txpoolOp.AddTxnReq{
+				Raw: &any.Any{
+					Value: stakeTxn.MarshalRLP(),
+				},
+				From: types.ZeroAddress.String(),
+			}
+		}
+
+		_, addErr := txpoolClient.AddTxn(context.Background(), msg)
+		if addErr != nil {
+			t.Fatalf("Unable to add txn, %v", addErr)
+		}
+	}
+
+	// Set up the blockchain listener to catch the added block event
+	blockNum := waitForBlock(t, srv, 1, 0)
+
+	block, blockErr := client.Eth().GetBlockByNumber(web3.BlockNumber(blockNum), true)
+	if blockErr != nil {
+		t.Fatalf("Unable to fetch block")
+	}
+
+	// Find how much the account paid for all the transactions in this block
+	paidFee := big.NewInt(0).Mul(gasPrice, big.NewInt(int64(block.GasUsed)))
+
+	// Check the balances
+	actualStakedBalance := getStakedBalance(senderAddr, client, t)
+	actualAccountBalance := getAccountBalance(senderAddr, client, t)
+	actualStakingAddrBalance := getAccountBalance(stakingContractAddr, client, t)
+
+	expStake := tests.EthToWei(1)
+
+	// Make sure the staked balance matches up
+	assert.Equalf(t,
+		expStake.String(),
+		actualStakedBalance.String(),
+		"Staked balance mismatch after stake / unstake exploit",
+	)
+
+	assert.Equalf(t,
+		(big.NewInt(0).Add(expStake, initialStakingAddrBalance)).String(),
+		actualStakingAddrBalance.String(),
+		"Staked address balance mismatch after stake / unstake exploit",
+	)
+
+	// Make sure the account balances match up
+
+	// expBalance = previousAccountBalance + stakeRefund - 1 ETH - block fees
+	expBalance := big.NewInt(0).Sub(big.NewInt(0).Add(defaultBalance, defaultBalance), oneEth)
+	expBalance = big.NewInt(0).Sub(expBalance, paidFee)
+
+	assert.Equalf(t,
+		expBalance.String(),
+		actualAccountBalance.String(),
+		"Account balance mismatch after unstake exploit",
+	)
+}
+
+func TestPoS_StakeUnstakeWithinSameBlock(t *testing.T) {
```

Comment:
Isn't the previous tests testing the staking/unstaking in the same block?

Also is there any reason the devInterval is 5 just in this test?

Reply:
The test scenario is different for this one, where it starts with a stake, then an unstake, whereas other start with an unstake then a stake. Initial params are also different.

No specific reason for lowering the dev interval to 5s, I was just wondering if it's enough time to wait, and forgot to make it uniform. Seems like 5s is enough for the dev interval, do you want me to change it for the other tests as well?